<div class="alert alert-info">
<font size="5"><b>Комментарий ревьюера</b></font>

Привет Станислав! Меня зовут Марат, и я буду твоим ревьюером. Спешу сообщить что все ключевые этапы в работе выполнены,  с задачей тебе удалось справиться. По поводу обращения - в IT сфере принято общаться на «ты» :) Но, если привычней на «вы», дай знать. Как ревьюера моя задача помочь тебе в развитии, дав хорошие советы. Я внимательно посмотрю твой код, ознакомлюсь с твоими выводами и оставлю комментарии. Где то могу предложить небольшие исправление в коде, но ненавязчиво. Где потребуются уточнения, я оставлю много наводящих вопросов. Они помогут тебя с поиском верного решения.

Все мои комментарии размечены по цветам, для лучшего восприятия. 
    
<div class="alert alert-success">Зеленым цветом и словом «Успех» отмечены особо удачные и элегантные решения, которыми ты можешь гордиться. </div>
        
<div class="alert alert-warning">Желтым и значком словом «Совет», помечены решения у которых есть альтернативные решения, более оптимальные. Ты можешь найти их сразу и доработать проект, или отложить это на потом, для будущих проектах. Проект будет принят и без их доработки. </div>
        
<div class="alert alert-danger"> Красным цветом и значком словом «Ошибка» помечу твои решения, на которые стоит обратить внимание прежде всего. После их доработки проект будет принят. </div>
        
Залог успеха - работа сообща, взаимное уважение и работа в диалоге. Поэтому, помечай свои ответные комментарии на мои реплики заметным цветом или курсивом, так мне будет легче их отслеживать. Пожалуйста, не изменяй и не удаляй мои комментарии. Все это поможет выполнить повторную проверку быстрей.

</div>

# Рекомендация тарифов

В вашем распоряжении данные о поведении клиентов, которые уже перешли на эти тарифы (из проекта курса «Статистический анализ данных»). Нужно построить модель для задачи классификации, которая выберет подходящий тариф. Предобработка данных не понадобится — вы её уже сделали.

Постройте модель с максимально большим значением *accuracy*. Чтобы сдать проект успешно, нужно довести долю правильных ответов по крайней мере до 0.75. Проверьте *accuracy* на тестовой выборке самостоятельно.

<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюера</b></font>

Успех:


Вступление в работу очень важно, так человек, который смотрит твой проект (и на работе в том числе) будет сразу введен в курс дела.





</div>

## Откройте и изучите файл

In [19]:
import pandas as pd 

from sklearn.tree import DecisionTreeClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import mean_squared_error
from sklearn.model_selection import train_test_split

from tqdm import tqdm

 <div class="alert alert-info">  Добавил from tqdm import tqdm (но что-то нигде не отображается) 

<div class="alert alert-warning">
<font size="5"><b>Комментарий ревьюераV2</b></font>



Совет 🤔:


Привет Станислав!
    
    

Потому что не дописал:
    
    
    for depth in range (1, 11):  -----> for depth in tqdm(range (1, 11)):    


In [55]:
df = pd.read_csv('/datasets/users_behavior.csv')

print(df.head(5))

   calls  minutes  messages   mb_used  is_ultra
0   40.0   311.90      83.0  19915.42         0
1   85.0   516.75      56.0  22696.96         0
2   77.0   467.66      86.0  21060.45         0
3  106.0   745.53      81.0   8437.39         1
4   66.0   418.74       1.0  14502.75         0


<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюера</b></font>

Успех:

Собираем все импорты в верхней части, чтобы легче было ориентироваться и добавлять новые по необходимости. 



<div class="alert alert-warning">

Совет:     
    


- Станислав, все-таки стоит импорт библиотек и выгрузку данных разнести в разные ячейки, так стилистически более правильно.


- кстати есть рекомендации PEP-8 при написании кода, в том числе и для импортов. Если интересно можешь почитать [тут](https://pythonworld.ru/osnovy/pep-8-rukovodstvo-po-napisaniyu-koda-na-python.html), в будущем пригодится )  



<div>

 <div class="alert alert-info">  V2 Сделал

<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюераV2</b></font>



Успех 👍:



👍




<div class="alert alert-warning">



Совет 🤔:

print не всегда нужен, вывод часто произойдет и так, а когда нужен, вместо print лучше использовать display, будет красивей





In [56]:
df.info() #Общая информация
# В задании указано что данные обработаны и предобработка данных не понадобится 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3214 entries, 0 to 3213
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   calls     3214 non-null   float64
 1   minutes   3214 non-null   float64
 2   messages  3214 non-null   float64
 3   mb_used   3214 non-null   float64
 4   is_ultra  3214 non-null   int64  
dtypes: float64(4), int64(1)
memory usage: 125.7 KB


<div class="alert alert-info"> Данные представлены в виде датасета о Звонках, Длительностью в Минутах, Количестве сообщений, Израсходованного трафика, Отношение к Тарифу Ультра. 

In [57]:
df['messages'] = df['messages'].astype('int') #Для наглядности меняем тип данных 
print(df.head(5))

   calls  minutes  messages   mb_used  is_ultra
0   40.0   311.90        83  19915.42         0
1   85.0   516.75        56  22696.96         0
2   77.0   467.66        86  21060.45         0
3  106.0   745.53        81   8437.39         1
4   66.0   418.74         1  14502.75         0


In [58]:
df.describe()

calls      minutes     messages       mb_used     is_ultra
count  3214.000000  3214.000000  3214.000000   3214.000000  3214.000000
mean     63.038892   438.208787    38.281269  17207.673836     0.306472
std      33.236368   234.569872    36.148326   7570.968246     0.461100
min       0.000000     0.000000     0.000000      0.000000     0.000000
25%      40.000000   274.575000     9.000000  12491.902500     0.000000
50%      62.000000   430.600000    30.000000  16943.235000     0.000000
75%      82.000000   571.927500    57.000000  21424.700000     1.000000
max     244.000000  1632.060000   224.000000  49745.730000     1.000000

<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюера</b></font>

Успех:

👍 Данные изучены.



<div class="alert alert-warning">



Совет: 


- .info (дает тебе информацию о пропусках и типах данных), а .describe (медианы, квантили, max, min итд) это такой джентльменский набор при первичной разведке данных (EDA).  



- Можно еще отдельно проверить датасет на сбалансированность классов в таргете. Это помогло бы решить бонусное задание сразу (Только не надо балансировать данные, это тема следующего проекта).




- Можно посмотреть корреляцию признаков. Знаешь что такое мультиколлинеарность, какие два типа проблем возникает, для каких моделей и какие варианты решения?




</div>


<div class="alert alert-info"> 
    
-1) сделал
    
-2) Не знаю, такого не проходил, возьму на заметку
    
-3) Тоже не знаю, в курсе этого не было. Я пока только по обучению, на остальное пока времени нет :( Основная работа есть к соажлению (


<div class="alert alert-warning">
<font size="5"><b>Комментарий ревьюераV2</b></font>


Совет 🤔:

  
    



- Дисбаланс - это тема следующего проекта.  Смысл очень простой, просто смотрим долю нулей и единичек в таргете. Как видишь нулей у нас 69% - это сразу решение бонусного задания. Именно accuracy 0,69 наша модель должна превзойти чтобы быть признанной адекватной






- Мультиколлинеарность. Если своими словами то Мультиколлинеарность - это высокая корреляция между признаками (у нас это calls-minutes) 


Она вызывает ряд проблем, и от неё можно избавиться, а можно не избавляться, потому что    


- Первый момент: мультиколлинеарность проблема линейных моделей - могут переобучаться.... теоретически. Но в логистической, и в линейной регрессиях реализована регуляризация, которая решает проблему мультиколлинеарности признаков. "Деревянные" модели (RF, DT) устойчивы к мультиколлинеарности, а кластеризация тоже почти не зависит от линейных зависимостей. Получается, избавляться от мультиколлинеарности чтобы не переобучаться не обязательно )) 
    
    
- Второй момент: интерпретируемость линейной модели (мы по коеффициентам судим как признаки влияют на таргет), тут  мультиколлинеарность надо убирать.


Так что получается что в нашем случае убирать мультиколлинеарность имеет смысл только если мы захотим посмотреть как конкретно признаки влияют на Таргет

   
И да, один их способов убрать  мультиколлинеарность это убрать один признак из каждой пары высокореллированных признаков.    


## Разбейте данные на выборки

In [11]:
df_train, df_valid_test = train_test_split(df, test_size=0.4, random_state=12345)

#Разбиваем на обучающую и валидо-тестовую V2

In [12]:
df_valid, df_test = train_test_split(df_valid_test, test_size=0.5, random_state=12345)
#Разбиваем на валидоционную и тестовую выборки пополам v2

In [38]:
train_features = df_train.drop(['is_ultra'], axis=1)
train_target = df_train['is_ultra']
valid_features = df_valid.drop(['is_ultra'], axis=1)
valid_target = df_valid['is_ultra']
test_features = df_test.drop(['is_ultra'], axis=1)
test_target = df_test['is_ultra']
#выделяем признаки и цель V2

<div class="alert alert-danger">
<font size="5"><b>Комментарий ревьюера</b></font>

Ошибка:

Информация в выборках не должна пересекаться, а ты два сплита сделал от df, в итоге получил явно завышенную метрику на валидации

<div class="alert alert-warning">



Совет: 


- Обрати внимание на аргумент stratify, он позволит сохранить изначальное распределение таргетов во всех новых датасетах.  Существующий дисбаланс никуда не денется, но в каждом датасете он будет одинаковым. [Почитать](https://pythonru.com/baza-znanij/sklearn-train-test-split) можно тут


- После разбиения лучше перестраховаться и использовать .shape, для контроля за корректностью разбиения




<div class="alert alert-info"> V2 Исправил, понял. Спасибо! 
     аргумент stratify пока не применял, пока сложно для внедрения
    

<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюераV2</b></font>



Успех 👍:



👍




Для первого сплита стратификацию можно было сделать так 
    
    
    stratify = df['is_ultra']
    
    
Для второго Догадайся сам )

In [41]:
df.shape

(3214, 5)

In [42]:
df_train.shape #60% от изначального фрейма


(1928, 5)

In [43]:
df_valid_test.shape  #40% от изначального фрейма и делим его пополам далее

(1286, 5)

In [44]:
df_valid.shape # OK!

(643, 5)

## Исследуйте модели

Модель RandomForestRegressor

In [21]:
%%time

best_accuracy = 0 
best_depth = 0 

for depth in range (1, 11):
    for est in range(5, 51, 5):
        for sample in range(2,5):
            model = RandomForestClassifier(random_state=12345,n_estimators=est, max_depth=depth, min_samples_leaf=sample) 
            model.fit(train_features, train_target)
            accuracy = model.score(valid_features, valid_target)
            if accuracy > best_accuracy: 
                best_accuracy = accuracy
                best_depth = depth
                best_est = est
                best_sample = sample
                           

print('Количество деревьев:', best_est, 'Максимальная глубина:', best_depth, 'мин. объектов в узле:', best_sample )
print('Качество', best_accuracy)

Количество деревьев: 35 Максимальная глубина: 6 мин. объектов в узле: 2
Качество 0.8087091757387247
CPU times: user 18.9 s, sys: 91.5 ms, total: 19 s
Wall time: 19 s




<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюера</b></font>

Успех:
   


 
По большему числу гиперпараметров перебор -   тем лучше метрика. Случайный лес мощная модель и по ней действительно стоит поперебирать несколько гиперпораметров. Здорово   




<div class="alert alert-warning">


Совет: 




- Когда что то долго крутиться, можно использовать  %%time - ставишь на самый вверх ячейки с кодом, время выполнения которого хочешь замерить, может не знаешь.  Быстрее не станет, но все будут видеть стоит ли ждать не отходя от ПК или можно сходить чаек поставить )) 
    
    Или tqdm, это ещё лучше, потому что он показывает интерактивно, на каком этапе расчетов мы находимся

    
    from tqdm import tqdm


    for n_estimators in tqdm(range(3,58)):

        ..........



(дальше мы будем уходить от неэффективных в питоне циклов к apply.  Там тоже есть аналог прогресс-баров:  .progress_apply)
    
И будет красиво )   
 



<div class="alert alert-info"> V2 вставил наверху from tqdm import tqdm  , но вроде не отображается

<div class="alert alert-warning">
<font size="5"><b>Комментарий ревьюераV2</b></font>



Совет 🤔:



Выше объяснил


Модель DecisionTreeClassifier

In [52]:
%%time
best_accuracy = 0 
best_depth = 0 

for depth in range(1,15):
    model = DecisionTreeClassifier(max_depth=depth, random_state=12345)
    model.fit(train_features, train_target)
    accuracy= model.score(valid_features, valid_target)
    if accuracy > best_accuracy: 
        best_accuracy = accuracy
        best_depth = depth       
               
    
print('Глубина дерева:', best_depth)
print('Качество', best_accuracy)

Глубина дерева: 3
Качество 0.7853810264385692
CPU times: user 95.1 ms, sys: 72 µs, total: 95.2 ms
Wall time: 93.2 ms


<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюера</b></font>

Успех:
 
Работа выполнена в соответствии с критериями: 



 - модель обучена на обучающем наборе
 - получена оценка качества на валидационном наборе
 - перебор гиперпараметров осуществляется в цикле





<div class="alert alert-warning">


Совет: 


- Можно добавить график, хорошо оформленный график украсит проект. Тут на графике можно посмотреть как при изменении max_depth меняется точность на train и validation датасете (для этого собираем наши метрики на валидации и train в списки, а затем по ним обычный plt.plot()) Такой способ используют чтобы лучше отследить переобучение, но для этого стоит взять пернебор по max_depth поглубже


- tree_plot можно построить, с max_depth равной 3 или 4 (глубже уже будет громоздко) - глянуть как DT сплиты делает, будет понимание как модель принимает решения: какие признаки  использует, на каких значениях делает разбиение вправо - влево 




</div>



<div class="alert alert-info">V2 Не совсем понял, Марат. Что за график? Можешь более развернуто написать вкратце, прям сейчас наверное не буду делать, но посмотрю дополнительно. 



<div class="alert alert-warning">
<font size="5"><b>Комментарий ревьюераV2</b></font>

Совет:

1. tree_plot
    
(схема)
    
    from sklearn import tree
    import matplotlib.pyplot as plt
    fig = plt.figure(figsize=(35,20))
    _ = tree.plot_tree(best_model_tree, 
                   feature_names=features.columns.values, 
                   class_names=['Smart','Ultra'],
                   filled=True,
                   fontsize = 15,
                   max_depth = 3     # Это чтобы прорисовывало 3 уровня, не глубже
                      )
    
    
    
2. График метрики  
    
    
Добавляешь (схема):
    
    
    predictions_train = model.predict(features_train)
    result_train = accuracy_score(target_train, predictions_train)

    
а теперь собираешь в цикле  result_train и result в списки (с помощью .append()), и по очереди подаешь эти списки в plt.plot()    
    
  


Модель LogisticRegression

In [46]:
%%time
model = LogisticRegression(solver='lbfgs')    
model.fit(train_features, train_target)  
accuracy = model.score(valid_features, valid_target)

print('Качество', accuracy)

Качество 0.7107309486780715
CPU times: user 21.6 ms, sys: 24 µs, total: 21.6 ms
Wall time: 20.1 ms



<div class="alert alert-warning">
<font size="5"><b>Комментарий ревьюера</b></font>

Совет: 

Если есть желание можешь ответить на вопросики ))    



- Как назвать ситуацию, когда на валидации при увеличении глубины дерева (модель DT), метрика качества стала снижаться? Как по твоему в это же время вела себя эта же метрика на train?


- Насколько знаю вы еще под капот моделям не заглядывали, но может знаешь почему обычно RF показывает более высокие результаты из выбранных?


- Мы решаем задачу классификации (а еще есть задача регрессии), а в названии модели с помощью которой мы решаем задачу "классификации" (Логистическая регрессия) есть слово "регрессия".  Нет ли тут парадокса? )
    

- Почему логистическая регрессия показывать на много более худшие результаты?  


</div>





<div class="alert alert-info">V2 1) Думаю ты про переобучении модели. На тренировочном наборе модель показывает себя хорошо, а на валидацинном плохо - значит модель переобучилась и лукавит на новых данных. 
    
2) Было вскольз в теории практикума - потому что RandomForest а) Это совокупность и среднее "голосование" деревьев" за результат б) использование randomstate= позволяет искать разные способы алгоритмов, добавляя случайности в последовательность алгоритмо, чтобы избежать переобучения
    
3) Навазание регрессия в данном случае ознает анализ и поиск зависимостей между данными
    
4) Логистическая регрессия хорошо подходит для задач когда имеется линейная разделимость классов. Но, если классы имеют сложную структуру и не могут быть разделены линейно, то логистическая регрессия может показать худшие результаты.

<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюераV2</b></font>


Успех:


Дополню чуть

1. Верно, переобучение. На train точность растет вплоть до 100%. 
    
    
2. Да, в RF обучается несколько деревьев, а решение принимается путем голосования. Каждое дерево в RF учится на своем датасете (используется техника бутстрап), таким образом достигается "независимость" этих самых деревьев, в итоге ошибка разброса снижается.

    
3. "Регрессия", потому что внутри нее скалярное произведение (как и в линейной регрессии), дающее на выходе вещественное число (от + бесконечности до - бесконечности), на это число вешают сигмойду, логарифм и получают значение от 0 до 1 (вероятность класса).   
    
    
4. Правильно. Логистическая регрессия плоха (и хороша с другой стороны) тем что это линейный классификатор: обьекты на классы разделяем гиперплоскостью (прямой если на плоскости), а в данном датасете видимо более сложные, нелинейные связи между признаками и таргетом.   А RF с DT с этим справляются. На других данных  LR могла оказаться лучше, например потому что RF не может экстраполировать. 


</div>

<div class="alert alert-info"> V2 Проверил три модели - RandomForest, DecisionTree и LogisticRegression. Лучшая по  accuracy = model.score(valid_features, valid_target) - стала RandomForest 0.808

## Проверьте модель на тестовой выборке

In [53]:
model = RandomForestClassifier(n_estimators=35, max_depth=6, min_samples_leaf=2, random_state=12345 )
model.fit(train_features, train_target)
accuracy = model.score(test_features, test_target)
print('Качество:', accuracy)

Качество: 0.7962674961119751


<div class="alert alert-info">V2  На тестовой выборке accuracy 0.79 , RandomForest при вычесленных параметрах выше n_estimators=35, max_depth=6, min_samples_leaf=2, random_state=12345 . То есть лучшая рекомендация тарифа для клиентов будет у этой модели.  



<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюераV2</b></font>

Успех:

Если студент получил на тесте accuraсy  выше 0,78, это считается отличным результатом. Тобой подбиралась лучшая комбинация не по одному гиперпараметру и вот он результат!






</div>


<div class="alert alert-danger">
<font size="5"><b>Комментарий ревьюера</b></font>




Ошибка:




К данному этапу мы должны подойти с одной выбранной моделью. Станислав, о логике использования датасетов:
    




1. На train мы обучаем
2. По валидации смотрим на результаты обучения (следим чтобы не было переобучения и/или делаем подбор гиперпараметров).  И выбираем лучшую модель. 
3. Тестовая (out-of_sample) у нас для финальной проверки, когда определена модель с конкретными гиперпараметрами. Использование тестовой выборки, чтобы еще раз посмотреть на модели - это нарушение логической цепочки моделирования. 

А в таком виде, может сложиться впечатление что мы на тестовой по прежнему что то выбираем, но выбор сделан, и если лучшая на валидации покажет на test результат хуже требуемого, мы начнем процесс моделирования сначала.
    
 [Вот](https://towardsdatascience.com/why-do-we-need-a-validation-set-in-addition-to-training-and-test-sets-5cf4a65550e0   ) тут можно дополнительно почитать.

<div class="alert alert-warning">
   
Совет: 

- Как попробовать улучшить результат, не мучаясь с кодом?! Смотри - мы сформировали train выборку на которой обучаем модель, а валидационную используем для поиска лучших значений гиперпараметров. Оке, нашли их. Так почему бы теперь модель с выбранными гиперпараметрами не обучить на тренировочной + валидационной выборке (только не нужно делать новые сплиты, используй pd.concat(), к примеру)?! Чем больше данных, тем лучше модель может обучиться. И эту дообученную модель проверим на тесте. Гаранитий что станет лучше нет, но ML это постоянные эксперименты.
</div>



In [50]:
model = RandomForestClassifier(n_estimators=35, max_depth=6, min_samples_leaf=2, random_state=12345 )
train_valid_df = pd.concat([df_train, df_valid], ignore_index=True)
train_valid_features = train_valid_df.drop(['is_ultra'], axis=1)
train_valid_target = train_valid_df['is_ultra']
model.fit(train_valid_features, train_valid_target)
accuracy = model.score(test_features, test_target)
print('Качество:', accuracy)

Качество: 0.7916018662519441


<div class="alert alert-info">V2 на модели без объеденения было - Качество: 0.7962674961119751 , стало Качество: 0.7916018662519441
    Лучше не стало, но логику понял! Спасибо! 
    
    Лишние модели в проверке удалил, тоже понял почему :) 

<div class="alert alert-warning">
<font size="5"><b>Комментарий ревьюераV2</b></font>



Совет 🤔:



Тут всегда присутствует Рандом, на этой тестовой выборке ухудшение, на десяти других будет улучшение.  Поэтому дообучать стоит, особенно если у нас такой маленький датасет как тут


## (бонус) Проверьте модели на адекватность

Добрый день! А где про это почитать на курсе? Самому изучить? Может попозже вернусь к этому, если сейчас не обязательно. Надо нагнать 



<div class="alert alert-warning">
<font size="5"><b>Комментарий ревьюера</b></font>

Совет:


Это если что может объяснить ревьюер )
    
    
Станислав, бонусное на то и бонусное, но можешь попробовать его решить с моей подсказкой.    

Ключевое это помнить  что у нас дисбаланс в таргете (нулей гораздно больше чем единиц) и четко понимать как считается accuracy.
В чем смысл проверки модели на адекватность  при дисбалансе? Вот смотри - обучили мы какую то модель, которая на test показывает 90% accuraсy. Шикарная модель скажем мы. А если в датасете у нас в метках таргета единичек 99,9% (допустим тренируем на данных где 1 это "кирпич упал на голову человеку когда он вышел из дома"), а нулей 0,1% ("не упал")?! Все также утверждаем что наша модель шикарная?! Нет, потому что всегда можно предсказывать что не упадает (единички) и accuracy будет равна 99,9%

Вот и тут мы сравниваем нашу обученную модель с примитивным прогнозом самого частого таргета. Какой у нас самый частый класс? Нули. Так посмотри accuracy_score target_test и ряда из одних нулей. Сколько будет?. 


    
    

При проверке на адекватность модели, рекомендую ознакомиться с [Dummy Models](https://scikit-learn.org/stable/modules/generated/sklearn.dummy.DummyClassifier.html), это удобная обертка для константных моделей и сделает тоже самое что я описал на словах

</div>

<div class="alert alert-warning">
<font size="5"><b>Комментарий ревьюераV2</b></font>

Совет:


Тут логика такая - у нас есть данные и мы видим что доля нулей в таргете около 69%, это значит что мы не проводя никакого анализа, моделирования итп итд, можем постоянно предсказывать 0 и получить точность около 69%. Вот эту accuracy в 69% мы и должны побить, иначе окажется что наша модель не лучше наивного прогноза самого частого класса.
      
    
С   DummyClassifier это выглядит  так:
    
    DummyClassifier(strategy='most_frequent', random_state=1)  
задаем константную модель, которая будет предсказывать по most_frequent - самый частый класс
    
    .fit(features_train, target_train)
тут модель смотри на самый частый класс в  target_train - типо "обучение"
    
    .predict(features_test)
    .score(features_test, target_test)
проогнозирует везде самый частый класс в features_test и считает score
    
Получаем тот же результат: acc около 69%.    



Сравниваем с accuracy нашей модели и делаем вывод



<div class="alert alert-info">
<font size="5"><b>Комментарий ревьюера</b></font>




Станислав, у тебя старательно выполненная работа, все четко, осмысленно. 
    
    
    


Я оставил небольшие советы и вопросики (если есть время и желание можешь воспользоваться/ответить): 




- обрати внимание на проверку сбалансированность классов в таргете в разделе EDA
- использовать stratify
- добавить график метрики    
- как можно улучшить результаты 



Замечания на будущее:
    


- Было бы замечательно, в выводе немножко о данных, работе с ними (пусть даже никакой работы и не было), моделировании и полученных результатах. Это важно потому что когда проект захочет посмотреть будущий работодатель (или начальник), у него может не быть времени на подробный разбор кода. Вероятнее всего он бегло просмотрит код, а из общего вывода захочет получить представление о всей работе.

 
    
Обязательное к исправлению:



- выборки не должны у нас пересекаться, иначе сразу нарушается логика моделирования

    
    
- к моменту проверки качества прогноза на тесте, приходим с уже одной выбранной моделью с лучшими гиперпараметрами выбранными на валидации

    
    
    
Жду исправлений, для принятия проекта. Если какие то вопросы, то сразу спрашивай ) 





</div>

<div class="alert alert-info"> Вывод о работе с данными: 
    В ходе анализа и работы с данными были применены три модели обучения с вычисляющмися в цикле гиперпараметрами моделей. RandomForest показал наилучшую эффективность accuracy 0.796 на тестовой выборке. LogisticRegression, DecisionTreeClassifier показали результаты хуже - 0.68 и 0.76 соответственно на тестовых выборках. Чтобы улучшать модели возможно потребуются более расширенные значения гиперпараметров и добавление новых признаков по клиентам. 


<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюераV2</b></font>

Успех 👍:

Общий вывод добавил

<div class="alert alert-info"> V2 Исправил выборки и подправил модели. Дописал вывод по работе с данными.
    
    
-использовать stratify
    
-добавить график метрики - пока тяжеловато для внедрения в работу. Отправляю пока так, позже постараюсь вернутся. 


<div class="alert alert-info">
<font size="5"><b>Комментарий ревьюераV2</b></font>

    
    
Значит остальные жёлтые советы используешь позже )  
    
    
Спасибо за работу!  

    
Красное исправлено,  желтые советы использованы, на  вопросы есть ответы  (это все было по желанию), значит стремишься развиваться, а желание и интерес это главное. Надеюсь мои советы и вопросики были полезны и в копилочку знаний упало что то новое, а проект стал лучше, и симпатичней.



Отличная работа Станислав. Желаю успехов в дальнейшей учебе!



## Чек-лист готовности проекта

Поставьте 'x' в выполненных пунктах. Далее нажмите Shift+Enter.

- [x] Jupyter Notebook открыт
- [ ] Весь код исполняется без ошибок
- [ ] Ячейки с кодом расположены в порядке исполнения
- [ ] Выполнено задание 1: данные загружены и изучены
- [ ] Выполнено задание 2: данные разбиты на три выборки
- [ ] Выполнено задание 3: проведено исследование моделей
    - [ ] Рассмотрено больше одной модели
    - [ ] Рассмотрено хотя бы 3 значения гипепараметров для какой-нибудь модели
    - [ ] Написаны выводы по результатам исследования
- [ ] Выполнено задание 3: Проведено тестирование
- [ ] Удалось достичь accuracy не меньше 0.75
